### PR TITLE
Support RangeResult where the Source provider can pre-seek to the start

### PR DESCRIPTION
**Pull request by halfninja**
Apologies that this does not yet include a Java API change or documentation - I will squash that in once someone's had a look to make sure this approach looks sane.

# Pull Request Checklist

* [x] Have you read [How to write the perfect pull request](https://github.com/blog/1943-how-to-write-the-perfect-pull-request)?
* [x] Have you read through the [contributor guidelines](https://www.playframework.com/contributing)?
* [x] Have you signed the [Lightbend CLA](https://www.lightbend.com/contribute/cla)?
* [x] Have you referenced any issues you're fixing using [commit message keywords](https://help.github.com/articles/closing-issues-using-keywords/)?
* [x] Have you added copyright headers to new files?
* [x] Have you checked that both Scala and Java APIs are updated?
* [ ] Have you updated the documentation for both Scala and Java sections?
* [x] Have you added tests for any changed functionality?

# Helpful things

## Purpose

When requesting the very end of a very large file, it can be inefficient to open a file at the start and drop N bytes. This is even more true for data backed by an HTTP API such as S3, where having to download the entire file to serve the last few bytes is extremely inefficient. This allows Source providers to seek directly to the start of the requested range.

## Background Context

Our application stores files in a Swift deployment, and needs to efficiently serve range requests.

By the time ByteStrings are being processed an HTTP response will have already begun, making it too late to implement preseeking at this stage. I've only added a new overload - existing method overloads remain the same.

I don't know what people's opinions are on having the function return a (Long, Source) tuple - would it be worth a case class for clarity? I went with passing just the Long offset around rather than a range object as those are all currently package private and I think exposing those would constitute a larger API change.

### CODE REVIEW

File: core/play/src/main/scala/play/api/mvc/RangeResult.scala [hunk 1]
```diff
@@ -447,6 +468,7 @@ object RangeResult {
         entityLength match {
           case Some(entityLen) =>
             if (entityLen > 0) {
+              val (_, source) = getSource(0L)
```
[the remark this answers is not in the document]
Comment (halfninja) [reply core/play/src/main/scala/play/api/mvc/RangeResult.scala]: Don't think we do - the first call is in the SatisfiableRangeSet case so wouldn't have occurred if we reached here

**Comment by halfninja** [comment]
> Hey @halfninja, this is looking good for me. :-)
> 
> I left some comments, and, as you said, this Java API counterpart is missing. I think it could also be interesting to add it to 2.8.0 highlights.
> 
> Thanks for taking the time to contribute here.

Thanks @marcospereira for your suggestions - I'll work on these and get the Java API and the documentation up to scratch.

**Comment by mkurz** [comment]
@halfninja Any chance you can finish this pull request? Would be nice to get this into the next release. Thanks!

**Comment by halfninja** [comment]
Yes, sorry, been trying to fit this round my regular work commitments.

**Comment by mkurz** [comment]
@halfninja How are you? Any chance you find time to finish this pull request? Would be a nice feature for Play's upcoming  2.8 release... :+1:

**Comment by halfninja** [comment]
Hello! 

@mkurz I've _finally_ added the Java API with tests. I haven't updated any documentation at the moment because currently the only mention of `RangeResult` is a simple use of `ofFile` in AssetsOverview, and I'm not sure if it's the best place to dunk people straight into source-generating functions, especially as the intended callers of the functions are likely to just be maintainers of libraries like [Alpakka](https://github.com/akka/alpakka). What do you reckon?

**Comment by mkurz** [comment]
> ... What do you reckon?
I pass this question on to @marcospereira :wink:

**Comment by marcospereira** [comment]
@halfninja I've submitted some changes here, hope you don't mind. We want this to be part of the next release candidate, that is why I add some commits myself.

Anyway, there is now documentation for range results, including pre-seeking. Besides that, your commits were squashed in a single one.

File: documentation/manual/working/javaGuide/main/http/JavaResponse.md [hunk 2]
```diff
@@ -43,12 +49,32 @@ To discard a Cookie previously stored on the web browser:
...
+
+@[range-result-input-stream](code/javaguide/http/JavaResponse.java)
+
+Or for an `Source`:
+
+@[range-result-source](code/javaguide/http/JavaResponse.java)
+
+When the request `Range` is not satisfiable, for example, if the range in the request's `Range` header field do not overlap the current extent of the selected resource, then a HTTP status `416` (Range Not Satisfiable) is returned.
+
+It is also possible to pre-seek for a specific position of the `Source` to more efficiently delivery range results. To do that, you can provide a function where the pre-seek happens:
```
Comment (halfninja) [inline documentation/manual/working/javaGuide/main/http/JavaResponse.md]: *deliver

File: documentation/manual/working/scalaGuide/main/http/ScalaResults.md [hunk 3]
```diff
@@ -68,3 +68,23 @@ Now if you are wondering how the `HTML` method works, here it is how it is defin
...
+
+@[range-result-input-stream](code/ScalaResults.scala)
+
+Or for an `Source`:
+
+@[range-result-source](code/ScalaResults.scala)
+
+When the request `Range` is not satisfiable, for example, if the range in the request's `Range` header field do not overlap the current extent of the selected resource, then a HTTP status `416` (Range Not Satisfiable) is returned.
+
+It is also possible to pre-seek for a specific position of the `Source` to more efficiently delivery range results. To do that, you can provide a function where the pre-seek happens:
```
Comment (halfninja) [inline documentation/manual/working/scalaGuide/main/http/ScalaResults.md]: *deliver

**Pull request merged**